ajax/kanban.php

```diff
@@ -190,9 +190,12 @@
     }
 } else if (($_POST['action'] ?? null) === 'move_item') {
     $checkParams(['card', 'column', 'position', 'kanban']);
-    /** @var Kanban|CommonDBTM $kanban */
+    /** @var CommonDBTM $kanban */
```

Comment:
```suggestion
    /** @var Kanban&CommonDBTM $kanban */
```

Reply:
I've tried, but it fails locally. Does it work on your system?

Reply:
See comment above, I guess method_exist works and is safe but it is not the best practice.

We can discuss it with @cedric-anne next week to be sure but the "clean" way to fix should be with an interface because PHP does not allow type hinting with traits .
-> We need an interface that define all methods found in the `Kanban` trait.
-> Any class that use the `Kanban` trait must declare that they implement this new interface
-> Any code that run on an "unknown" class that want to call kanban methods must check with instanceOf that the class implement the interface

Reply:
I guess the solution for cleanness is to make an `AbstractKanban` abstract class extending `CommonDBTM`, and safely type-hint over it.

Reply:
Exactly. Traits feel like a half implemented feature that make you *think* you have something even close to multiple inheritance except you dont, and you basically can't use traits on their own properly.

You have no way to know when a class uses a trait at runtime without reflection unless you don't care about parent classes.

PHPStan apparently doesn't properly support them even though I don't see any reason why it couldn't.

They basically just don't exist at runtime.

Reply:
> PHPStan apparently doesn't properly support them even though I don't see any reason why it couldn't.
They basically just don't exist at runtime.

This is on purpose: https://phpstan.org/writing-php-code/phpdoc-types#classes-and-interfaces

You can't typehint traits in PHP so it is impossible to validate that a given item has a trait method (`@var` would be cheating, the type has to be checked somewhere by actual code).

With that in mind it seems reasonable for PHPStan to behave this way, I guess you should blame PHP itself if you want "type hintable" traits (some languages have them).

Reply:
Traits are not meant for inheritance or typing, they're used only for composition.

Here, Kanban is used by 2 classes that extend `CommonDBTM`, and the trait provides the methods.
If you want typing, in both cases you will need a `KanbanInterface` that defines methods in the trait, but you might also need typing against `CommonDBTM` is valid in both classes.

The only way to have a proper type here is either Interface + Trait, or Abstract class that extends CommonDBTM and defines all the methods in the trait (so we can drop the trait entirely). There's no other solution for clean types. The interface can be created in both cases by the way, but it's not always necessary.

Reply:
I agree, the proper way to check that a class "implements" a trait is to have an interface that describe this trait, and then use `$this instanceof KanbanInterface` checks or `Common&KanbanInterface $param` type hinting.